vsphere/internal/helper/hostsystem/host_system_helper.go

```diff
@@ -116,11 +144,31 @@ func EnterMaintenanceMode(host *object.HostSystem, timeout int, evacuate bool) e
 		return err
 	}
 
-	return task.Wait(ctx)
+	err = task.Wait(ctx)
+	if err != nil {
+		return err
+	}
+	var to mo.Task
+	err = task.Properties(context.TODO(), task.Reference(), nil, &to)
+	if err != nil {
+		log.Printf("[DEBUG] Failed while getting task results: %s", err)
+		return err
+	}
+
+	if to.Info.State != "success" {
```

Comment:
Is there a non-success task state that can occur without an error?

Reply:
My understanding is that all vSphere API calls that return a task are asynchronous. If we get an error when we make the call it usually means that our parameters are wrong or that vsphere wasn't able to start doing the task. 

My expectation is that if something goes wrong between the time the API call returns a task ID and the time the task completes then it will be reflected in the task's Info structure.

(For example vsphere timed out while trying to talk to the host)

Reply:
You are correct on the vSphere API, but `govmomi` includes some helpers to wait for the task to complete and make sure it is successful. A few steps down the rabbit hole, and `task.Wait()` leads to [this](https://github.com/vmware/govmomi/blob/master/task/wait.go#L117) function which waits for the task to finish with either an "failure" or "success" state.

As long as the task is using `Wait`, you should be ok to trust that no error means it completed successfully.

Reply:
I had no idea :) ! I will give that a look, thanks!

Reply:
done